crates/flux-infer/src/fixpoint_encoding.rs

```diff
@@ -296,28 +299,45 @@ impl SortEncodingCtxt {
     }
 
     fn append_adt_decls(
+        &mut self,
         genv: GlobalEnv,
-        adt_sorts: FxIndexSet<DefId>,
         decls: &mut Vec<fixpoint::DataDecl>,
     ) -> QueryResult {
-        for (idx, adt_def_id) in adt_sorts.iter().enumerate() {
-            let adt_id = AdtId::from_usize(idx);
-            let adt_sort_def = genv.adt_sort_def_of(adt_def_id)?;
-            decls.push(fixpoint::DataDecl {
-                name: fixpoint::DataSort::Adt(adt_id),
-                vars: adt_sort_def.param_count(),
-                ctors: adt_sort_def
-                    .variants()
-                    .iter_enumerated()
-                    .map(|(idx, variant)| {
-                        debug_assert_eq!(variant.fields(), 0);
-                        fixpoint::DataCtor {
-                            name: fixpoint::Var::DataCtor(adt_id, idx),
-                            fields: vec![],
-                        }
-                    })
-                    .collect(),
-            });
+        // We iterate until we have processed all adt sorts because processing one adt sort may
+        // discover new adt sorts to process (e.g., if an adt field has an adt sort).
+        let mut done = FxHashSet::default();
+        while done.len() < self.adt_sorts.len() {
+            let adt_sorts = self.adt_sorts.clone();
+            for (idx, adt_def_id) in adt_sorts.iter().enumerate().skip(done.len()) {
+                if !done.insert(*adt_def_id) {
+                    continue;
+                }
+                let adt_id = AdtId::from_usize(idx);
```

Comment:
`adt_sort` is an `FxIndexSet` which is a set, but also preserves insertion order and can be indexed by position, so this can be simplified to

```rust
        let mut idx = 0;
        while let Some(adt_def_id) = self.adt_sort.get_index(idx) {
            let adt_id = AdtId::from_usize(idx);
            .
            .
            .
            idx += 1;
        }
```

Reply:
Although I wonder if there's a smarter way to traverse this, as they almost come out in reverse dependency order except for the initial set...